vavr/src/main/java/io/vavr/collection/Traversable.java

```diff
@@ -190,47 +183,101 @@
-     * List.of("apple", "pear").average()  // throws
+     * List.empty().average()                       // = None
+     * List.of(1, 2, 3).average()                   // = Some(2.0)
+     * List.of(1.0, 10e100, 2.0, -10e100).average() // = Some(0.75)
+     * List.of("apple", "pear").average()           // throws
+     * List.of(1.0, Double.NaN).average()           // NaN
      * </code>
      * </pre>
      *
+     * Please note that Java's {@link DoubleStream#average()} uses the
+     * <a href="https://en.wikipedia.org/wiki/Kahan_summation_algorithm">Kahan summation algorithm</a>
+     * (also known as compensated summation), which has known limitations.
+     * <p>
+     * Vavr uses Neumaier's modification of the Kahan algorithm, which yields better results.
+     *
+     * <pre>{@code
+     * // = OptionalDouble(0.0) (wrong)
+     * j.u.s.DoubleStream.of(1.0, 10e100, 2.0, -10e100).average()
+     *
+     * // = Some(0.75) (correct)
+     * List.of(1.0, 10e100, 2.0, -10e100).average()
+     * }</pre>
+     *
      * @return {@code Some(average)} or {@code None}, if there are no elements
-     * @throws UnsupportedOperationException if this elements are not numeric
+     * @throws ClassCastException if this elements are not numeric
      */
-    @SuppressWarnings({ "unchecked", "OptionalGetWithoutIsPresent" })
     default Option<Double> average() {
-        if (isEmpty()) {
+        return average(t -> ((Number) t).doubleValue());
+    }
+
+    /**
+     * Calculates the average of this elements using a function that maps this elements to {@code double} values.
+     * Returns {@code None} if this is empty, otherwise {@code Some(average)}.
+     * <p>
+     * Examples:
+     * <pre>{@code
+     * List.<Integer> empty().average(Number::doubleValue)             // = None
+     * List.of(1, 2, 3).average(Number::doubleValue)                   // = Some(2.0)
+     * List.of(1.0, 10e100, 2.0, -10e100).average(Number::doubleValue) // = Some(0.75)
+     * List.of(1.0, Double.NaN).average(Number::doubleValue)           // NaN
+     * }</pre>
+     *
+     * Please note that Java's {@link DoubleStream#average()} uses the
+     * <a href="https://en.wikipedia.org/wiki/Kahan_summation_algorithm">Kahan summation algorithm</a>
+     * (also known as compensated summation), which has known limitations.
+     * <p>
+     * Vavr uses Neumaier's modification of the Kahan algorithm, which yields better results.
+     *
+     * <pre>{@code
+     * // = OptionalDouble(0.0) (wrong)
+     * j.u.s.DoubleStream.of(1.0, 10e100, 2.0, -10e100).average()
+     *
+     * // = Some(0.75) (correct)
+     * List.of(1.0, 10e100, 2.0, -10e100).average(Number::doubleValue)
+     * }</pre>
+     *
+     * @param toDouble a function that converts elements to {@code double} values
+     * @return {@code Some(average)} or {@code None}, if there are no elements
+     */
+    default Option<Double> average(ToDoubleFunction<? super T> toDouble) {
```

Comment:
Gah! The dread `ToDoubleFunction`! I would LOVE to have a compiler switch that turns off primitives

Reply:
@chb0github The longer I play with sum, product, average, et al. the more I think you are right and we should remove the unsafe ones in 1.0. I like the `average(ToDoubleFunction)`.

`sum()` and `product()` are basically a `fold()`. However, we could provide also `double`-only versions `sum(ToDoubleFunction)` and `product(ToDoubleFunction)`. This way we would be safe.

I would keep `min()` and `max()` which assume that elements are comparable.